namecoin/src/main/java/org/libdohj/names/NameLookupLatestLevelDBTransactionCache.java

```diff
@@ -0,0 +1,410 @@
+import org.bitcoinj.core.StoredBlock;
+import org.bitcoinj.core.Transaction;
+import org.bitcoinj.core.TransactionOutput;
+import org.bitcoinj.core.VerificationException;
+import org.bitcoinj.core.listeners.NewBestBlockListener;
+import org.bitcoinj.core.listeners.ReorganizeListener;
+import org.bitcoinj.core.listeners.TransactionReceivedInBlockListener;
+import org.bitcoinj.script.Script;
+import org.bitcoinj.store.BlockStore;
+import org.bitcoinj.store.BlockStoreException;
+import org.bitcoinj.utils.Threading;
+
+import org.fusesource.leveldbjni.*;
+import org.iq80.leveldb.*;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.common.collect.HashMultimap;
+import com.google.common.collect.Multimaps;
+import com.google.common.collect.SetMultimap;
+
+import java.io.*;
+import java.nio.*;
+import java.util.Date;
+import java.util.EnumSet;
+import java.util.List;
+
+// TODO: dead blocks' name transactions are stored in memory indefinitely.  We should probably fix that, although doing so will slow down processing of reorgs.
+
+public class NameLookupLatestLevelDBTransactionCache implements NameLookupLatest, NewBestBlockListener, ReorganizeListener, TransactionReceivedInBlockListener {
+    
+    
+    protected static final byte[] CHAIN_HEAD_KEY = "Head".getBytes();
+    protected static final byte[] HEIGHT_KEY = "Height".getBytes();
+    
+    protected BlockChain chain;
+    protected BlockStore store;
+    protected PeerGroup peerGroup;
+    
+    protected Context context;
+    protected NetworkParameters params;
+    protected File path;
+    
+    protected DB db;
+    
+    protected SetMultimap<Sha256Hash, Transaction> pendingBlockTransactions = Multimaps.synchronizedSetMultimap(HashMultimap.<Sha256Hash, Transaction>create());
+    
+    protected Logger log = LoggerFactory.getLogger(NameLookupLatestLevelDBTransactionCache.class);
+    
+    public NameLookupLatestLevelDBTransactionCache (Context context, File directory, BlockChain chain, BlockStore store, PeerGroup peerGroup) throws Exception {
+        this(context, directory, JniDBFactory.factory, chain, store, peerGroup);
+    }
+    
+    public NameLookupLatestLevelDBTransactionCache (Context context, File directory, DBFactory dbFactory, BlockChain chain, BlockStore store, PeerGroup peerGroup) throws Exception {
+        this.chain = chain;
+        this.store = store;
+        this.peerGroup = peerGroup;
+        
+        this.context = context;
+        this.params = context.getParams();
+        
+        this.path = directory;
+        Options options = new Options();
+        options.createIfMissing();
+        
+        try {
+            tryOpen(directory, dbFactory, options);
+        } catch (IOException e) {
+            try {
+                dbFactory.repair(directory, options);
+                tryOpen(directory, dbFactory, options);
+            } catch (IOException e1) {
+                throw new Exception(e1);
```

Comment:
Can you just do `throw e1;` please, and mark this method as throwing `IOException` please? Means it's easier for calling code to break down the different error types that can occur.